# **Daffa Asyqar Ahmad Khalisheka**

# **1103200034**

# **Machine Learning Class**

# **Telkom University**

Kode ini menggunakan library Google Colab untuk mengakses Google Drive. Dengan menggunakan fungsi `drive.mount('/content/drive')`, Google Drive dapat di-"mount" (dipasang) pada direktori `/content/drive` di lingkungan Google Colab. Hal ini memungkinkan pengguna untuk mengakses dan bekerja dengan file yang ada di Google Drive mereka langsung dari notebook Google Colab tersebut.

In [23]:
# Import library Google Colab untuk mengakses Google Drive
from google.colab import drive

# Mount (pasang) Google Drive pada direktori /content/drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


Kode tersebut adalah implementasi sederhana untuk membangun, melatih, dan mengevaluasi sebuah model jaringan saraf menggunakan TensorFlow di lingkungan Google Colab. Berikut penjelasan dari setiap baris kode:

1. **Import Library**: Mengimpor library yang diperlukan untuk analisis data dan pembangunan model jaringan saraf, seperti Pandas untuk manipulasi data, NumPy untuk operasi numerik, train_test_split untuk membagi data menjadi set pelatihan dan pengujian, StandardScaler untuk penskalaan fitur, dan TensorFlow untuk membangun dan melatih model jaringan saraf.

```python
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
```

2. **Load Data**: Kode ini tidak menyertakan langkah-langkah untuk memuat data, yang diasumsikan sebelumnya. Penting untuk memuat data yang akan digunakan dalam model. Data tersebut kemungkinan besar berupa dataframe Pandas.

3. **Data Preprocessing**: Langkah ini mencakup pemisahan data menjadi fitur (X) dan target (y), serta pembagian data menjadi set pelatihan dan pengujian menggunakan `train_test_split`. Selanjutnya, fitur dinormalisasi menggunakan `StandardScaler`.

4. **Build Neural Network Model**: Membangun model jaringan saraf menggunakan Sequential API dari TensorFlow. Model ini memiliki satu lapisan input dan satu lapisan output dengan beberapa lapisan tersembunyi di antaranya.

5. **Compile Model**: Mengkompilasi model dengan mengtentukan optimizer, fungsi kerugian (loss function), dan metrik evaluasi yang akan digunakan selama pelatihan.

6. **Training Model**: Melatih model pada data pelatihan dengan menentukan jumlah epoche dan ukuran batch.

7. **Evaluate Model**: Menggunakan set pengujian untuk mengevaluasi kinerja model yang telah dilatih.

Catatan: Kode tersebut tidak menyertakan detail data, struktur model, atau konfigurasi pelatihan yang lebih spesifik. Oleh karena itu, penting untuk memuat data yang sesuai dan menyesuaikan konfigurasi model sesuai dengan kebutuhan aplikasi.

In [58]:
# Import library yang diperlukan
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

Kode tersebut digunakan untuk membaca file CSV yang berisi data cuaca ke dalam sebuah DataFrame menggunakan pustaka Pandas di lingkungan Google Colab. Langkah-langkahnya adalah sebagai berikut:

1. `import pandas as pd`: Mengimpor pustaka Pandas dengan alias `pd`.

2. `data = pd.read_csv('/content/drive/MyDrive/Dataset/Weather_Data.csv')`: Membaca file CSV yang berada di path '/content/drive/MyDrive/Dataset/Weather_Data.csv' dan menyimpannya ke dalam variabel `data`. File CSV tersebut berisi data cuaca.

Penjelasan tambahan:
- `pd`: Alias untuk pustaka Pandas, mempermudah penggunaan fungsi dan metode dari Pandas.
- `read_csv()`: Fungsi Pandas untuk membaca file CSV dan mengonversinya menjadi DataFrame.
- `'/content/drive/MyDrive/Dataset/Weather_Data.csv'`: Lokasi file CSV yang akan dibaca. Path tersebut sesuai dengan struktur direktori Google Colab.
- `data`: Variabel yang menampung DataFrame hasil pembacaan file CSV, sehingga dapat digunakan untuk analisis data lebih lanjut.

In [24]:
#Membaca file CSV yang berisi data ke dalam sebuah DataFrame
data = pd.read_csv('/content/drive/MyDrive/Dataset/Weather_Data.csv')

Kode tersebut adalah perintah untuk menampilkan lima baris pertama dari dataset yang disimpan dalam variabel 'data'. Fungsi 'head()' digunakan untuk itu, dan kode ini umumnya digunakan ketika Anda ingin dengan cepat melihat struktur dan konten awal dari dataset yang sedang Anda kerjakan. Kode ini sering digunakan dalam lingkungan Google Colab, yang merupakan platform penelitian dan pengembangan berbasis cloud yang menggunakan notebook Jupyter.

In [25]:
#Menampilkan isi dari dataset
data.head()

Date  Temp9am  Temp3pm  MinTemp  MaxTemp  Rainfall RainToday  \
0  01-02-13     20.7     20.9     19.5     22.4      15.6       Yes   
1  02-02-13     22.4     24.8     19.5     25.6       6.0       Yes   
2  03-02-13     23.5     23.0     21.6     24.5       6.6       Yes   
3  04-02-13     21.4     20.9     20.2     22.8      18.8       Yes   
4  05-02-13     22.5     25.5     19.7     25.7      77.4       Yes   

   Evaporation  Sunshine WindGustDir  ...  WindDir9am WindDir3pm WindSpeed9am  \
0          6.2       0.0           W  ...           S        SSW           17   
1          3.4       2.7           W  ...           W          E            9   
2          2.4       0.1           W  ...         ESE        ESE           17   
3          2.2       0.0           W  ...         NNE          E           22   
4          4.8       0.0           W  ...         NNE          W           11   

   WindSpeed3pm  Humidity9am  Humidity3pm  Pressure9am  Pressure3pm  Cloud9am  \
0            20           92           84       1017.6       1017.4         8   
1            13           83           73       1017.9       1016.4         7   
2             2           88           86       1016.7       1015.6         7   
3            20           83           90       1014.2       1011.8         8   
4             6           88           74       1008.3       1004.8         8   

   Cloud3pm  
0         8  
1         7  
2         8  
3         8  
4         8  

[5 rows x 21 columns]

Output di atas adalah tampilan lima baris pertama dari dataset yang terdiri dari 21 kolom. Setiap baris mewakili data cuaca pada tanggal tertentu. Berikut adalah penjelasan singkat untuk setiap kolom:

1. **Date**: Tanggal pengamatan cuaca.
2. **Temp9am**: Suhu pada pukul 9 pagi.
3. **Temp3pm**: Suhu pada pukul 3 sore.
4. **MinTemp**: Suhu minimum pada hari tersebut.
5. **MaxTemp**: Suhu maksimum pada hari tersebut.
6. **Rainfall**: Jumlah curah hujan dalam milimeter.
7. **RainToday**: Apakah hujan terjadi hari itu (Yes/No).
8. **Evaporation**: Penguapan kelas A (milimeter) dalam 24 jam hingga pukul 9 pagi.
9. **Sunshine**: Jumlah jam sinar matahari dalam sehari.
10. **WindGustDir**: Arah angin puncak kecepatan.
11. ... (dan seterusnya): Kolom-kolom lainnya termasuk arah angin pada pukul 9 pagi dan 3 sore, kecepatan angin pada pukul 9 pagi dan 3 sore, kelembaban pada pukul 9 pagi dan 3 sore, tekanan atmosfer pada pukul 9 pagi dan 3 sore, serta kondisi awan pada pukul 9 pagi dan 3 sore.

Informasi ini memberikan gambaran awal tentang variabel-variabel yang terdapat dalam dataset dan kondisi cuaca pada lima hari pertama dalam data tersebut.

Kode tersebut digunakan untuk menampilkan fitur atau kolom-kolom yang ada dalam suatu dataset. Di dalam kode tersebut, `data.columns` mengacu pada atribut columns dari objek DataFrame yang disebut `data`. DataFrame ini adalah struktur data tabular yang digunakan dalam library pandas, yang umumnya digunakan untuk manipulasi dan analisis data di Python.

Jadi, `print(data.columns)` akan mencetak daftar nama kolom-kolom yang ada dalam dataset yang disimpan dalam variabel `data`. Ini berguna untuk memahami struktur dataset dan mengidentifikasi variabel-variabel atau fitur-fitur yang tersedia untuk analisis lebih lanjut.

Kode ini umumnya digunakan di lingkungan seperti Google Colab, yang merupakan lingkungan penulisan kode berbasis web yang berjalan di atas server Google, memungkinkan untuk melakukan analisis data menggunakan Python di cloud.

In [26]:
# Tampilkan fitur dalam dataset
print(data.columns)

Index(['Date', 'Temp9am', 'Temp3pm', 'MinTemp', 'MaxTemp', 'Rainfall',
       'RainToday', 'Evaporation', 'Sunshine', 'WindGustDir', 'WindGustSpeed',
       'WindDir9am', 'WindDir3pm', 'WindSpeed9am', 'WindSpeed3pm',
       'Humidity9am', 'Humidity3pm', 'Pressure9am', 'Pressure3pm', 'Cloud9am',
       'Cloud3pm'],
      dtype='object')


Output di atas merupakan daftar nama kolom atau fitur-fitur yang ada dalam dataset. Setiap nama kolom menggambarkan jenis data atau informasi yang disimpan di dalamnya. Berikut adalah penjelasan singkat untuk setiap kolom:

1. **Date**: Tanggal pengamatan cuaca.
2. **Temp9am**: Suhu pada pukul 9 pagi.
3. **Temp3pm**: Suhu pada pukul 3 sore.
4. **MinTemp**: Suhu minimum dalam sehari.
5. **MaxTemp**: Suhu maksimum dalam sehari.
6. **Rainfall**: Jumlah curah hujan.
7. **RainToday**: Apakah hujan terjadi hari ini (Yes/No).
8. **Evaporation**: Penguapan.
9. **Sunshine**: Durasi sinar matahari.
10. **WindGustDir**: Arah angin puncak kecepatan.
11. **WindGustSpeed**: Kecepatan angin puncak.
12. **WindDir9am**: Arah angin pada pukul 9 pagi.
13. **WindDir3pm**: Arah angin pada pukul 3 sore.
14. **WindSpeed9am**: Kecepatan angin pada pukul 9 pagi.
15. **WindSpeed3pm**: Kecepatan angin pada pukul 3 sore.
16. **Humidity9am**: Kelembaban pada pukul 9 pagi.
17. **Humidity3pm**: Kelembaban pada pukul 3 sore.
18. **Pressure9am**: Tekanan udara pada pukul 9 pagi.
19. **Pressure3pm**: Tekanan udara pada pukul 3 sore.
20. **Cloud9am**: Tutupan awan pada pukul 9 pagi.
21. **Cloud3pm**: Tutupan awan pada pukul 3 sore.

Kolom-kolom ini mengandung informasi terkait cuaca yang dapat digunakan untuk analisis dan pemodelan prediktif terkait kondisi cuaca di waktu dan lokasi tertentu.

Kode tersebut digunakan untuk menampilkan informasi tentang objek 'data'. Ini dapat memberikan pemahaman lebih lanjut tentang struktur dan karakteristik data yang ada. Pada kode tersebut, diasumsikan bahwa 'data' adalah suatu objek yang memiliki metode `info()`, yang umumnya digunakan pada platform seperti Google Colab yang menggunakan bahasa pemrograman Python dengan library pandas untuk manipulasi dan analisis data.

Penjelasan singkat:
1. `data`: Nama objek atau variabel yang berisi data yang ingin ditampilkan informasinya.
2. `.info()`: Metode dari objek 'data' yang digunakan untuk menampilkan informasi tentang struktur data seperti jumlah baris, jumlah kolom, tipe data setiap kolom, dan jumlah nilai non-null pada setiap kolom.

Keterangan tambahan:
- Kode ini dapat membantu dalam pemahaman awal tentang data yang digunakan, termasuk apakah terdapat nilai null, tipe data kolom, dan jumlah total baris dan kolom.
- Output dari metode `info()` biasanya berisi ringkasan informasi yang mencakup nama kolom, tipe data, dan jumlah non-null values untuk masing-masing kolom.

Pastikan untuk menjalankan kode di lingkungan yang sesuai dengan bahasa pemrograman dan library yang digunakan, seperti Python dan pandas di Google Colab.

In [27]:
# Ini adalah kode untuk menampilkan informasi tentang objek 'data'
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3271 entries, 0 to 3270
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Date           3271 non-null   object 
 1   Temp9am        3271 non-null   float64
 2   Temp3pm        3271 non-null   float64
 3   MinTemp        3271 non-null   float64
 4   MaxTemp        3271 non-null   float64
 5   Rainfall       3271 non-null   float64
 6   RainToday      3271 non-null   object 
 7   Evaporation    3271 non-null   float64
 8   Sunshine       3271 non-null   float64
 9   WindGustDir    3271 non-null   object 
 10  WindGustSpeed  3271 non-null   int64  
 11  WindDir9am     3271 non-null   object 
 12  WindDir3pm     3271 non-null   object 
 13  WindSpeed9am   3271 non-null   int64  
 14  WindSpeed3pm   3271 non-null   int64  
 15  Humidity9am    3271 non-null   int64  
 16  Humidity3pm    3271 non-null   int64  
 17  Pressure9am    3271 non-null   float64
 18  Pressure

Output tersebut adalah hasil dari pemanggilan metode `info()` pada suatu objek DataFrame dengan nama 'data'. Berikut adalah penjelasan dari setiap bagian output:

1. **Kelas DataFrame:**
   - Jenis: `pandas.core.frame.DataFrame`
   - Deskripsi: Menunjukkan bahwa objek 'data' adalah suatu DataFrame, yang merupakan struktur data tabular dua dimensi pada pandas.

2. **RangeIndex:**
   - Jenis: `RangeIndex`
   - Ukuran: 3271 entries, dari 0 sampai 3270
   - Deskripsi: Indeks default untuk DataFrame, dimulai dari 0 hingga 3270 dengan total 3271 entries.

3. **Data columns (total 21 columns):**
   - Deskripsi: Menunjukkan bahwa DataFrame 'data' memiliki 21 kolom.

4. **Detail Kolom:**
   - **Column**: Nama setiap kolom pada DataFrame.
   - **Non-Null Count**: Jumlah nilai non-null pada setiap kolom.
   - **Dtype**: Tipe data dari setiap kolom.

5. **Tipe Data Kolom (dtypes):**
   - Terdapat tiga jenis tipe data pada DataFrame:
     - **float64**: Tipe data untuk kolom yang berisi nilai desimal.
     - **int64**: Tipe data untuk kolom yang berisi nilai bilangan bulat.
     - **object**: Tipe data untuk kolom yang berisi nilai objek atau teks.

6. **Memory Usage:**
   - Ukuran: 536.8+ KB
   - Deskripsi: Menunjukkan perkiraan penggunaan memori oleh DataFrame 'data'.

Ringkasan:
- Data memiliki 21 kolom dan 3271 baris.
- Tidak ada nilai null pada setiap kolom.
- Tipe data meliputi float64, int64, dan object.
- Output ini memberikan gambaran umum tentang struktur dan sifat data yang sedang digunakan.

Kode tersebut menghapus baris dari suatu DataFrame yang memiliki nilai null. DataFrame adalah struktur data tabular dua dimensi yang umumnya digunakan dalam analisis data menggunakan library seperti pandas di Python. `data` dalam kode tersebut adalah nama DataFrame yang sedang dioperasikan.

Pada baris pertama, `data.dropna()` digunakan untuk menghapus semua baris yang mengandung setidaknya satu nilai null dalam DataFrame `data`. Fungsi `dropna()` adalah metode bawaan dari pandas yang digunakan untuk menghilangkan baris atau kolom yang berisi nilai null.

Kode tersebut efektif membersihkan DataFrame dari baris yang tidak memiliki nilai lengkap atau tidak ada data yang hilang. Penting untuk memperhatikan bahwa perubahan ini dilakukan pada DataFrame yang ada di memori, dan jika Anda ingin menyimpan perubahan tersebut, Anda perlu mengonfirmasinya dengan menyimpan DataFrame yang telah diubah ke variabel baru atau pada variabel yang sama.

In [48]:
# Menghapus baris yang memiliki nilai null
data= data.dropna()

Kode tersebut mendefinisikan sebuah list yang disebut `features` yang berisi beberapa string yang mewakili nama-nama fitur. Setiap string dalam list ini mencerminkan jenis data atau informasi yang terkait dengan kondisi cuaca, seperti suhu pada pukul 9 pagi (`Temp9am`), suhu pada pukul 3 sore (`Temp3pm`), suhu minimum (`MinTemp`), suhu maksimum (`MaxTemp`), curah hujan (`Rainfall`), evaporasi (`Evaporation`), jumlah sinar matahari (`Sunshine`), kecepatan angin puncak (`WindGustSpeed`), kecepatan angin pada pukul 9 pagi (`WindSpeed9am`), kecepatan angin pada pukul 3 sore (`WindSpeed3pm`), kelembaban pada pukul 9 pagi (`Humidity9am`), kelembaban pada pukul 3 sore (`Humidity3pm`), tekanan udara pada pukul 9 pagi (`Pressure9am`), tekanan udara pada pukul 3 sore (`Pressure3pm`), jumlah awan pada pukul 9 pagi (`Cloud9am`), dan jumlah awan pada pukul 3 sore (`Cloud3pm`).

Kode ini kemungkinan besar digunakan untuk pemrosesan atau analisis data cuaca, di mana fitur-fitur ini akan digunakan sebagai variabel input atau atribut dalam suatu model atau analisis statistik. List `features` ini dapat digunakan di berbagai macam konteks, seperti dalam pengembangan model prediksi cuaca, analisis tren cuaca, atau tugas-tugas lain yang melibatkan data cuaca.

In [49]:
# Memilih fitur yang akan digunakan
features = ['Temp9am', 'Temp3pm', 'MinTemp', 'MaxTemp', 'Rainfall', 'Evaporation', 'Sunshine',
            'WindGustSpeed', 'WindSpeed9am', 'WindSpeed3pm', 'Humidity9am', 'Humidity3pm',
            'Pressure9am', 'Pressure3pm', 'Cloud9am', 'Cloud3pm']

Kode tersebut mendefinisikan variabel dengan nama "target" dan memberinya nilai string 'Rainfall'. Variabel ini kemungkinan besar digunakan sebagai suatu target atau tujuan dalam konteks program atau analisis data yang menggunakan kode tersebut. Penjelasan lebih lanjut tentang penggunaan variabel "target" ini dapat ditemukan di bagian-bagian selanjutnya dari kode atau program yang lebih luas. Perlu diingat bahwa informasi yang diberikan disesuaikan dengan konteks yang diberikan dan tidak memberikan detail tentang bagaimana variabel ini digunakan dalam konteks yang lebih besar. Selain itu, tidak ada informasi tentang lingkungan pengembangan yang digunakan, tetapi disebutkan bahwa kode ini ditulis atau dijalankan di Google Colab, yang merupakan platform berbasis cloud untuk pengembangan dan eksekusi kode Python.

In [50]:
# Memilih target
target = 'Rainfall'

Kode tersebut digunakan untuk memisahkan dataset menjadi dua bagian utama: satu untuk fitur (X) dan satu untuk target (y). Pemisahan ini umumnya dilakukan dalam konteks pembelajaran mesin, di mana X akan berisi variabel-variabel independen atau fitur-fitur yang akan digunakan untuk membuat prediksi, dan y akan berisi variabel dependen atau target yang ingin diprediksi.

Berikut adalah penjelasan untuk setiap baris kode:

1. `X = data[features]`: Dari dataset `data`, kita mengambil subset fitur-fitur yang akan digunakan untuk pembelajaran mesin. Variabel `features` berisi nama-nama fitur yang ingin dimasukkan ke dalam X. Jadi, X menjadi subset dari data yang hanya berisi kolom-kolom yang sesuai dengan fitur-fitur yang telah ditentukan.

2. `y = data[target]`: Dari dataset `data` juga, kita mengambil kolom yang sesuai dengan target yang ingin diprediksi. Variabel `target` berisi nama target yang ingin diprediksi. Sehingga, y menjadi subset dari data yang hanya berisi kolom target.

Pemisahan ini memungkinkan kita untuk dengan mudah mengakses fitur-fitur yang akan digunakan dalam pembelajaran mesin serta target yang akan diprediksi. Dengan cara ini, kita dapat menyusun data ke dalam bentuk yang sesuai untuk proses pelatihan dan evaluasi model pembelajaran mesin.

In [51]:
# Memisahkan fitur dan target
X = data[features]
y = data[target]


Kode tersebut menggunakan fungsi `train_test_split` dari suatu pustaka (sepertinya scikit-learn) untuk membagi dataset menjadi data latih (X_train, y_train) dan data uji (X_test, y_test). Proses pembagian dilakukan dengan menggunakan parameter `test_size=0.2`, yang berarti data uji akan menyusun 20% dari keseluruhan dataset. Selain itu, parameter `random_state=42` digunakan untuk menghasilkan pembagian yang konsisten setiap kali kode dijalankan, sehingga hasilnya dapat direproduksi dengan mudah.

Pada umumnya, X adalah matriks fitur atau input, dan y adalah vektor target atau output yang sesuai. Proses ini umumnya digunakan dalam pembelajaran mesin untuk mengevaluasi kinerja model pada data yang tidak terlihat selama pelatihan.

In [52]:
# Memisahkan data menjadi data latih dan data uji
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


Kode tersebut merupakan implementasi dari sebuah model neural network menggunakan library Keras di Google Colab. Model ini memiliki tiga lapisan (layers) dengan jumlah neuron tertentu pada setiap lapisan. Berikut adalah penjelasan rinci dari setiap bagian kode:

1. `model = Sequential()`: Membuat model neural network dengan urutan lapisan sekuensial.

2. `model.add(Dense(64, input_dim=len(features), activation='relu'))`: Menambahkan lapisan Dense pertama dengan 64 neuron, mengambil input dengan jumlah dimensi yang sesuai dengan panjang fitur (features). Fungsi aktivasi yang digunakan adalah ReLU.

3. `model.add(Dense(32, activation='relu'))`: Menambahkan lapisan Dense kedua dengan 32 neuron dan menggunakan fungsi aktivasi ReLU.

4. `model.add(Dense(1, activation='linear'))`: Menambahkan lapisan Dense ketiga dengan satu neuron untuk output dan menggunakan fungsi aktivasi linear.

Penjelasan singkat mengenai setiap lapisan:
   - Lapisan pertama (`Dense(64, input_dim=len(features), activation='relu')`): Lapisan input dengan 64 neuron, menerima input sesuai dengan panjang fitur, dan menggunakan fungsi aktivasi ReLU.
   - Lapisan kedua (`Dense(32, activation='relu')`): Lapisan tersembunyi dengan 32 neuron dan fungsi aktivasi ReLU.
   - Lapisan ketiga (`Dense(1, activation='linear')`): Lapisan output dengan satu neuron dan fungsi aktivasi linear.

Model ini dirancang untuk tugas regresi, di mana outputnya berupa nilai kontinu. Fungsi aktivasi ReLU digunakan di lapisan tersembunyi untuk memperkenalkan non-linearitas, sedangkan fungsi aktivasi linear pada lapisan output sesuai untuk tugas regresi.

In [53]:
# Membangun model neural network
model = Sequential()
model.add(Dense(64, input_dim=len(features), activation='relu'))
model.add(Dense(32, activation='relu'))
model.add(Dense(1, activation='linear'))

Kode tersebut merupakan bagian dari proses kompilasi model dalam pembelajaran mesin menggunakan library Keras, yang mana Keras merupakan antarmuka tingkat tinggi untuk membangun dan melatih model neural networks.

Pada kode di atas, model neural network dikompilasi dengan menggunaan optimizer 'adam' dan fungsi kerugian (loss function) 'mean_squared_error'. Optimizer dan fungsi kerugian ini adalah dua komponen kunci dalam proses pelatihan model.

- Optimizer 'adam' digunakan untuk mengoptimalkan model dengan meminimalkan nilai fungsi kerugian. 'Adam' merupakan salah satu optimizer populer yang menggabungkan konsep dari metode-metode optimasi lainnya.
  
- Fungsi kerugian 'mean_squared_error' (MSE) digunakan untuk mengukur seberapa baik model memprediksi nilai target dengan membandingkan nilai prediksi dengan nilai sebenarnya dan menghitung rata-rata kuadrat dari selisihnya. Fungsi ini cocok untuk masalah regresi di mana tujuannya adalah untuk memprediksi nilai numerik.

Dengan mengkompilasi model dengan optimizer dan fungsi kerugian yang sesuai, model siap untuk dilatih pada data pelatihan menggunakan metode backpropagation dan algoritma optimasi yang telah dipilih.

In [54]:
# Kompilasi model
model.compile(optimizer='adam', loss='mean_squared_error')

Kode tersebut merupakan implementasi pelatihan (training) model menggunakan metode pembelajaran mesin (machine learning) dengan menggunakan library atau framework tertentu, mungkin dalam konteks ini menggunakan TensorFlow atau Keras. Berikut adalah penjelasan dari setiap bagian kode:

1. `model.fit`: Ini adalah metode untuk melatih model. Melibatkan proses iteratif di mana model diperbarui berdasarkan data latih (training data) untuk meningkatkan performanya.

2. `X_train` dan `y_train`: Merupakan data latih (features dan labels) yang digunakan untuk melatih model.

3. `epochs=100`: Menunjukkan berapa kali seluruh dataset akan digunakan untuk melatih model. Dalam konteks ini, model akan dilatih sebanyak 100 kali.

4. `batch_size=32`: Menentukan jumlah sampel data yang digunakan dalam satu iterasi. Dalam hal ini, model akan diperbarui setiap kali telah melihat 32 sampel data.

5. `validation_data=(X_test, y_test)`: Data validasi digunakan untuk mengevaluasi model pada setiap epoch. Ini membantu memantau apakah model overfitting atau tidak.

Selanjutnya, secara keseluruhan, kode tersebut mengindikasikan bahwa model akan dilatih menggunakan data latih `X_train` dan `y_train` sebanyak 100 epoch, dengan ukuran batch sebesar 32, dan akan dievaluasi menggunakan data validasi `X_test` dan `y_test` setiap epoch. Ini adalah bagian dari proses pengembangan dan penyesuaian model untuk mencapai kinerja yang baik pada data yang belum pernah dilihat sebelumnya (data validasi).

In [56]:
# Melatih model
model.fit(X_train, y_train, epochs=100, batch_size=32, validation_data=(X_test, y_test))

Epoch 1/100
82/82 [==============================] - 1s 6ms/step - loss: 0.1005 - val_loss: 0.0404
Epoch 2/100
82/82 [==============================] - 0s 5ms/step - loss: 0.1080 - val_loss: 0.0647
Epoch 3/100
82/82 [==============================] - 0s 5ms/step - loss: 0.0690 - val_loss: 0.0392
Epoch 4/100
82/82 [==============================] - 0s 5ms/step - loss: 0.1234 - val_loss: 0.2254
Epoch 5/100
82/82 [==============================] - 0s 5ms/step - loss: 0.2301 - val_loss: 0.0565
Epoch 6/100
82/82 [==============================] - 0s 5ms/step - loss: 0.0890 - val_loss: 0.1676
Epoch 7/100
82/82 [==============================] - 0s 4ms/step - loss: 0.0964 - val_loss: 1.1215
Epoch 8/100
82/82 [==============================] - 0s 3ms/step - loss: 0.3871 - val_loss: 0.0275
Epoch 9/100
82/82 [==============================] - 0s 3ms/step - loss: 0.0683 - val_loss: 0.0342
Epoch 10/100
82/82 [==============================] - 0s 3ms/step - loss: 0.1103 - val_loss: 0.0292
Epoch 11/

Output tersebut merupakan hasil dari pelatihan (training) model dengan 100 epoch. Setiap epoch menunjukkan nilai loss pada data pelatihan (`loss`) dan data validasi (`val_loss`). Berikut adalah penjelasan dari setiap baris output:

- `Epoch 1/100`: Ini menunjukkan bahwa model sedang berada pada epoch pertama dari total 100 epoch yang telah ditentukan.
- `82/82 [==============================]`: Ini memberikan informasi tentang progress pelatihan. Jumlah 82 mengindikasikan iterasi atau batch yang telah diproses, dan [==============================] adalah tanda visual yang menunjukkan seberapa banyak pelatihan telah selesai.
- `- 1s 6ms/step - loss: 0.1005 - val_loss: 0.0404`: Informasi ini menyajikan metrik pelatihan. `loss: 0.1005` adalah nilai loss pada data pelatihan pada epoch tersebut, dan `val_loss: 0.0404` adalah nilai loss pada data validasi pada epoch tersebut.
- `Epoch 2/100`: Ini menunjukkan bahwa model telah memasuki epoch kedua dari total 100 epoch.
- `82/82 [==============================] - 0s 5ms/step - loss: 0.1080 - val_loss: 0.0647`: Seperti sebelumnya, ini memberikan informasi tentang progress dan metrik pelatihan pada epoch kedua.
- Output serupa untuk setiap epoch selanjutnya, hingga epoch ke-100.

Setelah pelatihan selesai, output terakhir menunjukkan bahwa model telah selesai dilatih dengan nilai loss terakhir pada data pelatihan sekitar 0.1690 dan nilai loss pada data validasi sekitar 0.0100. Itu ditunjukkan oleh bagian `Epoch 100/100 - ... - loss: 0.1690 - val_loss: 0.0100`. Model yang telah dilatih kemudian dapat digunakan untuk membuat prediksi pada data baru.

Kode ini dimaksudkan untuk membuat prediksi menggunakan model yang telah dilatih sebelumnya untuk memprediksi curah hujan berdasarkan data cuaca. Berikut adalah penjelasan langkah demi langkah dari kode tersebut:

1. **Import Library**: Memasukkan library yang diperlukan, dalam hal ini, numpy untuk manipulasi array dan model machine learning, serta StandardScaler dari scikit-learn untuk normalisasi data.

```python
import numpy as np
from sklearn.preprocessing import StandardScaler
```

2. **Data Baru**: Membuat array numpy yang merepresentasikan data baru yang akan diprediksi. Data ini memiliki 16 fitur.

```python
new_data = np.array([[-0.5, 0.3, -1.0, 1.5, 0.0, 2.0, 8.0, 30.0, 10.0, 15.0, 70.0, 40.0, 1010.0, 1005.0, 5.0, 4.0]])
```

3. **Objek Scaler Baru**: Membuat objek StandardScaler baru yang akan digunakan untuk normalisasi data baru.

```python
new_scaler = StandardScaler()
```

4. **Pelatihan Scaler**: Melatih scaler baru dengan menggunakan data pelatihan (X_train).

```python
new_scaler.fit(X_train)
```

5. **Normalisasi Data Baru**: Melakukan normalisasi terhadap data baru menggunakan scaler yang telah dilatih.

```python
scaled_new_data = new_scaler.transform(new_data)
```

6. **Prediksi**: Menggunakan model yang telah dilatih sebelumnya untuk membuat prediksi terhadap data baru yang sudah dinormalisasi.

```python
prediction = model.predict(scaled_new_data)
```

7. **Output**: Menampilkan hasil prediksi.

```python
print(f'Predicted Rainfall: {prediction[0][0]}')
```

Dengan menggunakan langkah-langkah tersebut, kode ini mengilustrasikan proses mempersiapkan data baru, normalisasi data, dan akhirnya membuat prediksi menggunakan model machine learning yang telah dilatih sebelumnya.

In [60]:
# Contoh prediksi untuk data baru
new_data = np.array([[-0.5, 0.3, -1.0, 1.5, 0.0, 2.0, 8.0, 30.0, 10.0, 15.0, 70.0, 40.0, 1010.0, 1005.0, 5.0, 4.0]])

# Membuat objek scaler baru
new_scaler = StandardScaler()

# Melatih scaler baru dengan data pelatihan
new_scaler.fit(X_train)

# Normalisasi data baru
scaled_new_data = new_scaler.transform(new_data)

# Prediksi menggunakan model
prediction = model.predict(scaled_new_data)
print(f'Predicted Rainfall: {prediction[0][0]}')


/usr/local/lib/python3.10/dist-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(


1/1 [==============================] - 0s 240ms/step
Predicted Rainfall: 0.742160439491272


Output tersebut terdiri dari dua bagian: pesan peringatan (warning) yang berasal dari scikit-learn dan hasil prediksi curah hujan.

1. **Pesan Peringatan (Warning)**:

   ```
   /usr/local/lib/python3.10/dist-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
     warnings.warn(
   ```

   Pesan ini memberikan peringatan kepada pengguna bahwa data yang digunakan untuk normalisasi (X_train) tidak memiliki nama fitur yang valid. Dalam konteks ini, StandardScaler dapat diinstruksikan dengan nama fitur untuk setiap kolom pada data, namun, karena data yang digunakan untuk pelatihan (X_train) tidak memiliki nama fitur yang valid, peringatan ini dihasilkan. Peringatan ini sendiri tidak mempengaruhi proses prediksi secara langsung, tetapi dapat diabaikan jika nama fitur tidak kritis dalam analisis.

2. **Hasil Prediksi**:

   ```
   1/1 [==============================] - 0s 240ms/step
   Predicted Rainfall: 0.742160439491272
   ```

   Bagian ini menunjukkan hasil prediksi dari model terhadap data baru setelah normalisasi. Dalam kasus ini, hasil prediksi curah hujan adalah sekitar 0.742. Nilai ini sesuai dengan apa yang dicetak dalam kode:

   ```python
   print(f'Predicted Rainfall: {prediction[0][0]}')
   ```

   Hasil ini menunjukkan jumlah curah hujan yang diprediksi oleh model untuk data baru yang diberikan setelah melalui proses normalisasi.